raft/src/main/java/org/apache/kafka/raft/KafkaRaftLog.java

```diff
@@ -0,0 +1,843 @@
+          follower will be unable to append it since (X - M) < (X).
+         */
+        long baseOffset = read(snapshotId.offset(), Isolation.COMMITTED).startOffsetMetadata.offset();
+        if (snapshotId.offset() != baseOffset) {
+            throw new IllegalArgumentException(
+                    "Cannot create snapshot at offset (" + snapshotId.offset() + ") because it is not batch aligned. " +
+                    "The batch containing the requested offset has a base offset of (" + baseOffset + ")"
+            );
+        }
+        return createNewSnapshotUnchecked(snapshotId);
+    }
+
+    @Override
+    public Optional<RawSnapshotWriter> createNewSnapshotUnchecked(OffsetAndEpoch snapshotId) {
+        boolean containsSnapshotId;
+        synchronized (snapshots) {
+            containsSnapshotId = snapshots.containsKey(snapshotId);
+        }
+
+        if (containsSnapshotId) {
+            return Optional.empty();
+        } else {
+            return Optional.of(
+                    new NotifyingRawSnapshotWriter(
+                            FileRawSnapshotWriter.create(log.dir().toPath(), snapshotId),
+                            this::onSnapshotFrozen
+                    )
+            );
+        }
+    }
+
+    @Override
+    public Optional<RawSnapshotReader> readSnapshot(OffsetAndEpoch snapshotId) {
+        synchronized (snapshots) {
+            Optional<FileRawSnapshotReader> reader = snapshots.get(snapshotId);
+            if (reader == null) {
+                return Optional.empty();
+            } else if (reader.isPresent()) {
+                return Optional.of(reader.get());
+            } else {
+                // Snapshot exists but has never been read before
+                try {
+                    FileRawSnapshotReader fileReader = FileRawSnapshotReader.open(log.dir().toPath(), snapshotId);
+                    snapshots.put(snapshotId, Optional.of(fileReader));
+                    return Optional.of(fileReader);
+                } catch (UncheckedIOException e) {
+                    // Snapshot doesn't exist in the data dir; remove
+                    Path path = Snapshots.snapshotPath(log.dir().toPath(), snapshotId);
+                    logger.warn("Couldn't read {}; expected to find snapshot file {}", snapshotId, path);
+                    snapshots.remove(snapshotId);
+                    return Optional.empty();
+                }
+            }
+        }
+    }
+
+    @Override
+    public Optional<RawSnapshotReader> latestSnapshot() {
+        synchronized (snapshots) {
+            return latestSnapshotId().flatMap(this::readSnapshot);
+        }
+    }
+
+    @Override
+    public Optional<OffsetAndEpoch> latestSnapshotId() {
+        synchronized (snapshots) {
+            try {
+                OffsetAndEpoch epoch = snapshots.lastKey();
+                return epoch == null ? Optional.empty() : Optional.of(epoch);
+            } catch (NoSuchElementException e) {
```

Comment:
How about changing this NoSuchElementException catching code to `if snapshots.isEmpty()`

Reply:
Good idea